azure/Guidelines.md

```diff
@@ -0,0 +1,243 @@
+
+### Specifying the version in Azure
+
+The Microsoft API guidelines give two options for how services and clients communicate the version: a url segment and a query parameter. Azure services **MUST** use the api-version query parameter. For example:
+
+```
+GET http://blobstore.azure.com/foo.com/acct1/c1/blob2?api-version=1.0
+PUT http://blobstore.azure.com/foo.com/acct1/c1/b2?api-version=2014-12-07
+POST http://blobstore.azure.com/foo.com/acct1/c1/b2?api-version=2015-12-07
+```
+
+### Breaking changes in Azure
+
+A breaking change is any change in the API that may cause client or service code making the API call to fail. Obvious examples of such a change are the removal of an endpoint, adding or removing a required field or changing the format of the body (from XML to JSON for example).
+
+Even though we recommend clients ignore new fields, there are many libraries and clients that fail when new fields are introduced. Azure services **MUST** update the version number of their API even when adding optional fields. In fact, servers should be as strict as possible. Ignoring a field can result in the API accepting content that containered a typo or an element at the wrong level of nesting. If this missing field changes the semantics (for example, we have seen cases where security settings were misplaced and ignored, leaving the resources more exposed than intended) this can be a huge and hard to discover error.
+
+At a high level, changes to the contract of an API constitute a breaking change. Changes that impact backwards compatibility of an API is also considered a breaking change. Teams MAY define backwards compatibility as their business needs require. For example, Azure defines the addition of a new JSON field in a response to be not backwards compatible. Anything that would violate the Principle of Least Astonishment is considered a breaking change in Azure. Below are some concrete examples of what constitutes a breaking change. In the below breaking change scenarios, the API version must be changed.
+
+#### Existing property is removed
+
+If a property called `foo` that was present in v1 of the API needs to be removed, it must be done in a newer API version.
+
+#### New property added to response
+
+If a new property/field is added to the response of an API, the GET-PUT pipeline will be broken. Consider the case where a customer updates the value of a new property "A" from the Azure portal. Another customer does a GET of this resource using the SDK. The SDK will ignore the property since it does not understand it. From the SDK, the customer does a PUT using the model that was returned from the GET. This will overwrite the change made by the first customer from the portal.
```

Comment:
This seems to be increasingly at-odds with standard web development practices. I'm not sure how long Azure will be able to hold the line on this.

Reply:
It's certainly a topic that can be visited during the update meetings that are planned.

Reply:
The Graph SDKs are able to avoid this problem because unknown properties retrieved from the server are round tripped.  This follows the guidelines defined by the [Atom publishing protocol](https://bitworking.org/projects/atom/rfc5023#rfc.section.9.3).  It is worth mentioning that Microsoft Graph does not make extensive use of PUT.

Reply:
This is a consequence of https://github.com/microsoft/api-guidelines/blob/vNext/Guidelines.md#61-ignore-rule

Reply:
And please note that we've had instances where a service would reject a call with a read-only property that a client didn't know, but still tried to round-trip. The requirement for client behavior has to be in lock step with how a service is to treat unknown or read-only properties.